proto/errors.go

```diff
@@ -73,16 +73,30 @@ var (
 type KafkaError struct {
 	errno   int16
 	message string
+	// An additional message explaining cases where the client returns this error internally
+	// instead of directly from a Kafka broker.
+	syntheticMessage string
 }
 
 func (err *KafkaError) Error() string {
+	if err.syntheticMessage != "" {
+		fmt.Sprintf("%s (%d %s)", err.message, err.errno, err.syntheticMessage)
+	}
 	return fmt.Sprintf("%s (%d)", err.message, err.errno)
 }
 
 func (err *KafkaError) Errno() int {
 	return int(err.errno)
 }
 
+func (err *KafkaError) SyntheticError() string {
+	return err.syntheticMessage
+}
+
+func (err *KafkaError) NewSynthetic(message string) *KafkaError {
+	return &KafkaError{err.errno, err.message, message}
```

Comment:
Hmm I didn't really think this through.

A lot of places rely on being able to compare the pointers -- i.e., `if err == proto.ErrUnknownTopicOrPartition` is making sure they're the exact objects but `NewSynthetic` will return a new object.

There's probably a couple possible fixes, what do you think?